lib/ziti-tunnel-cbs/ziti_tunnel_ctrl.c

```diff
@@ -57,6 +57,70 @@ const ziti_tunnel_ctrl* ziti_tunnel_init_cmd(uv_loop_t *loop, tunneler_context t
     return &CMD_CTX.ctrl;
 }
 
+
+int ziti_dump_to_log_cb(void* stringsBuilder, const char *fmt,  ...) {
+    static char line[4096];
+
+    va_list vargs;
+    va_start(vargs, fmt);
+    vsnprintf(line, sizeof(line), fmt, vargs);
+    va_end(vargs);
+
+    if (strlen(stringsBuilder) + strlen(line) > 8192) {
+        return -1;
+    }
+    // write/append to the buffer
+    strncat(stringsBuilder, line, sizeof(line));
+    return 0;
+}
+
+void ziti_dump_to_log(void *ctx) {
+    char* buffer;
+    buffer = malloc(8192*sizeof(char));
+    buffer[0] = 0;
+    //actually invoke ziti_dump here
+    ziti_dump(ctx, ziti_dump_to_log_cb, buffer);
+    ZITI_LOG(INFO, "ziti dump to log %s", buffer);
+    free(buffer);
+}
+
+int ziti_dump_to_file_cb(void* fp, const char *fmt,  ...) {
+    static char line[4096];
+
+    va_list vargs;
+    va_start(vargs, fmt);
+    vsnprintf(line, sizeof(line), fmt, vargs);
+    va_end(vargs);
+
+    // write/append to file
+    fputs(line, fp);
+
+    return 0;
+}
+
+void ziti_dump_to_file(void *ctx, char* outputFile) {
+    FILE *fp;
+    fp = fopen(outputFile, "a+");
+    if(fp == NULL)
+    {
+        ZITI_LOG(ERROR, "ziti dump to file failed. Unable to Read / Write / Create File");
+        return;
+    }
+    uv_timeval64_t dump_time;
+    uv_gettimeofday(&dump_time);
+
+    char time_str[32];
+    struct tm* start_tm = gmtime(&dump_time.tv_sec);
+    strftime(time_str, sizeof(time_str), "%FT%T", start_tm);
+
+    fprintf(fp, "Ziti Dump starting: %s\n",time_str);
+
+    //actually invoke ziti_dump here
+    ziti_dump(ctx, ziti_dump_to_file_cb, fp);
+    fflush(fp);
+    fclose(fp);
```

Comment:
Similar to freeing the buffer in `ziti_dump_to_log`, here I think the file needs to remain open until `ziti_dump_to_file_cb` is invoked and completed. The invocation is asynchronous to this function.

Reply:
I misunderstood the way ziti_dump is using the printer function. It's entirely synchronous, so the file and memory resources can be cleaned up here.